### PR TITLE
feature/cluster_rollingupgrade_when_certificate_renewal

### PR DESCRIPTION
**Pull request by PLsergent**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no|
| Deprecations?   | no
| License         | Apache 2.0

### What's in this PR?
This PR aims at creating a new field in the struct NodeState called CertificateExpireDate, and then automate the Nifi cluster rolling upgrade when the certificate is renewed (handle by the cert manager).

The idea is that we will give the field CertificateExpireDate the value of the current certificate renewal date, during each reconcile we will check if the CertificateExpireDate field has the same value as the current certificate renewal date, if no, we will do a rolling upgrade of the cluster and then assign the new renewal date to each node.

### Why?
When the certificate handle by the cert manager was renewed it was necessary to restart the Nifi nodes manually.

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](docs/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)
- [ ] Append changelog with changes

### To Do

- [ ] Rebase master

### CODE REVIEW

File: api/v1alpha1/nificluster_types.go [hunk 1]
```diff
@@ -491,6 +491,8 @@ type NifiClusterStatus struct {
 	RootProcessGroupId string `json:"rootProcessGroupId,omitempty"`
 	// PrometheusReportingTask contains the status of the prometheus reporting task managed by the operator
 	PrometheusReportingTask PrometheusReportingTaskStatus `json:"prometheusReportingTask,omitempty"`
+	// Certificate revision status contains the number of revision of the certificate (renewal)
+	ClusterRevision int `json:"revision,omitempty"`
```
Comment (juldrixx) [inline api/v1alpha1/nificluster_types.go]:
```suggestion
	CertificateRevision int `json:"certificateRevision,omitempty"`
```

File: controllers/nificluster_controller.go [hunk 2]
```diff
@@ -256,6 +259,7 @@ func (r *NifiClusterReconciler) checkFinalizers(ctx context.Context,
 		// user finalizations are done before it does its final cleanup
 		interval := util.GetRequeueInterval(r.RequeueIntervals["CLUSTER_TASK_NOT_READY_REQUEUE_INTERVAL"]/3, r.RequeueOffset)
 		r.Log.Info("Tearing down any PKI resources for the nificluster")
+		// PL
```
Comment (juldrixx) [inline controllers/nificluster_controller.go]:
```suggestion
```

File: controllers/nificluster_controller.go [hunk 3]
```diff
@@ -82,6 +83,8 @@ func (r *NifiClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 
 	// Fetch the NifiCluster instance
 	instance := &v1alpha1.NifiCluster{}
+
+	// pass a certificate instead of instance PL
```
Comment (juldrixx) [inline controllers/nificluster_controller.go]:
```suggestion
```

File: pkg/pki/certmanagerpki/certmanager_pki.go [hunk 4]
```diff
@@ -20,9 +20,33 @@ import (
...
+		// c.cluster.Status.ClusterRevision++
+		return true
+	}
+	return false
+}
+
 func (c *certManager) FinalizePKI(ctx context.Context, logger logr.Logger) error {
 	logger.Info("Removing cert-manager certificates and secrets")
 
+	// Logic get certmanager info PL
```
Comment (juldrixx) [inline pkg/pki/certmanagerpki/certmanager_pki.go]:
```suggestion
```

File: pkg/pki/certmanagerpki/certmanager_pki.go [hunk 5]
```diff
@@ -48,6 +72,7 @@ func (c *certManager) FinalizePKI(ctx context.Context, logger logr.Logger) error
 			// Delete the certificates first so we don't accidentally recreate the
 			// secret after it gets deleted
 			cert := &certv1.Certificate{}
+			// HERE PL
```
Comment (juldrixx) [inline pkg/pki/certmanagerpki/certmanager_pki.go]:
```suggestion
```

**Review by juldrixx — CHANGES_REQUESTED** [review]

**Comment by juldrixx** [comment]
Et au passage, la PR c'est pas sur ton projet à toi qu'il fallait la faire mais sur le projet source 😆

**Comment by juldrixx** [comment]
Pour le status sur la révision, selon moi il faut le mettre à jour à 2 moments. Le première fois quand la valeur n'est pas mise. Et une deuxième fois quand le rollingUpgrade est terminé.

**Comment by PLsergent** [comment]
Oui je voulais d'abord faire une PR sur mon projet avant de la faire du le vrai repo 🤒

**Pull request closed without merging**